### DISCUSSION TITLE
"inline" or non-"fullscreen" apps

### DISCUSSION THREAD

**Discussion by sullyj3** [jtdaugherty/brick#480] [Q&A] [answered]
Many interactive TUI apps clear the terminal and run in an experience that is sort of separate from the shell. This is how brick operates by default.
However, some TUI applications don't clear the screen, instead operating inline in the shell, rendering an interactive UI below at the current command prompt. Note that I do mean interactive programs, and not just CLI ones. The example that comes to mind is `fzf`, which looks like this:
![inline tui](https://github.com/jtdaugherty/brick/assets/3953669/e18c1cee-b172-4a58-a93a-d6b316ec8c79)

Does brick support this style of "inline tui"?

**Comment by sullyj3** [comment]
Having done some more research, it seems like the concept I'm getting at is something called the "alternate screen". I assume brick (or I guess the underlying vty) uses this by default? I will do some grepping... hm, no salient results for "alternate".

**Reply by jtdaugherty** [reply, marked as answer]
Hi @sullyj3. At the moment, `vty` (and thus `brick`) does not support any non-fullscreen operation. I don't know what it would take to support that so I couldn't say anything about what it would take to implement. (Or what the API considerations would need to be for both `vty` and `brick`.)

**Reply by sullyj3** [reply]
Thanks!

**Reply by sullyj3** [reply]
Collecting info in case it becomes useful in future:

Relevant code appears to be [here in vty](https://github.com/jtdaugherty/vty/blob/72fc3dc45c79c0e34c2bfb389d366b0137ecf845/src/Graphics/Vty/Output/TerminfoBased.hs#L176C29-L176C29)
```haskell
, reserveDisplay = do
    -- If there is no support for smcup: Clear the screen
    -- and then move the mouse to the home position to
    -- approximate the behavior.
    maybeSendCap smcup []
    sendCap clearScreen []
```

Call stack at that point is 
```
mkVty -> outputForConfig -> XTermColor.reserveTerminal -> TerminfoBased.reserveTerminal
-- or possibly
mkVty -> outputForConfig -> TerminfoBased.reserveTerminal
```

`smcup` seems to be terminfo for "switch to the alternate screen".

**Reply by jtdaugherty** [reply]
Also, for posterity, bear in mind that (as of now) `vty` is undergoing a great deal of refactoring to move the above code into a new `vty-unix` package that probably be released in the next couple of months. So if anyone digs into this, that'll be the place to look at that point.
